lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
+  for (EquationInstanceOp equationOp : instanceOps) {
+    llvm::SmallVector<EquationInstanceOp> &group =
+        instancesByTemplate.getOrInsertDefault(equationOp.getTemplate());
+    group.push_back(equationOp);
+  }
+
+  for (auto [_, instanceOps] : instancesByTemplate) {
+    if (haveSameInductionRanges(instanceOps)) {
+      filteredInstanceOps.push_back(instanceOps.front());
+    }
+  }
+  llvm::stable_sort(filteredInstanceOps);
+  return filteredInstanceOps;
+}
+
+/// Get all call operations in the model with their associated iteration spaces.
+llvm::SmallVector<CallWithIndices> collectCallOps(ModelOp modelOp) {
   llvm::SmallVector<EquationInstanceOp> dynamicEquationOps;
   modelOp.collectMainEquations(dynamicEquationOps);
 
-  llvm::DenseSet<EquationTemplateOp> visitedTemplateOps;
-  for (EquationInstanceOp equationOp : dynamicEquationOps) {
-    EquationTemplateOp templateOp = equationOp.getTemplate();
-    if (!templateOp.getInductionVariables().empty() ||
-        visitedTemplateOps.contains(templateOp)) {
-      continue;
-    }
-    visitedTemplateOps.insert(templateOp);
-    templateOp->walk([&](CallOp callOp) {
+  llvm::SmallVector<CallWithIndices> callOps;
+  for (EquationInstanceOp equationOp : filterInstanceOps(dynamicEquationOps)) {
+    equationOp.getTemplate()->walk([&](CallOp callOp) {
       // Skip functions with tensor results
       if (llvm::any_of(callOp.getResultTypes(), [](mlir::Type type) {
             return type.isa<mlir::TensorType>();
           })) {
         return;
       }
-      callOps.push_back(callOp);
+
+      std::optional<MultidimensionalRange> indices;
+      if (std::optional<MultidimensionalRangeAttr> indicesAttr =
+              equationOp.getIndices()) {
+        indices = indicesAttr.value().getValue();
+      }
+      callOps.push_back({callOp, indices});
     });
   }
+  return callOps;
 }
 
 /// Partition the list of call operations into groups given by
 /// EquationExpressionOpInterface::isEquivalent
-void buildCallEquivalenceGroups(
-    llvm::SmallVectorImpl<CallOp> &callOps,
-    llvm::SmallVectorImpl<llvm::SmallVector<CallOp>> &callEquivalenceGroups) {
-  mlir::SymbolTableCollection symbolTableCollection;
-  llvm::SmallVector<llvm::SmallVector<CallOp>> tmpCallEquivalenceGroups;
-
-  for (CallOp callOp : callOps) {
-    auto callExpression =
-        mlir::cast<EquationExpressionOpInterface>(callOp.getOperation());
-
-    llvm::SmallVector<CallOp> *equivalenceGroup = find_if(
-        tmpCallEquivalenceGroups, [&](llvm::SmallVector<CallOp> &group) {
+llvm::SmallVector<llvm::SmallVector<CallWithIndices>>
+buildEquivalenceGroups(const llvm::SmallVectorImpl<CallWithIndices> &callOps) {
+  llvm::SmallVector<llvm::SmallVector<CallWithIndices>> callEquivalenceGroups;
+  for (CallWithIndices callOp : callOps) {
+    auto callExpression = mlir::dyn_cast<EquationExpressionOpInterface>(
+        callOp.first.getOperation());
+
+    mlir::SymbolTableCollection symbolTableCollection;
```

Comment:
Do not recreate a `SymbolTableCollection` every time. This should be create only once within the pass.